cmd/dep/init.go

```diff
@@ -137,12 +99,15 @@ func (cmd *initCommand) Run(ctx *dep.Ctx, args []string) error {
 	if ctx.Verbose {
 		ctx.Out.Println("Getting direct dependencies...")
 	}
-	pkgT, directDeps, err := getDirectDependencies(sm, p)
+
+	// If this errors, the next call will too; don't bother handling it twice.
```

Comment:
This seems brittle.

Since `GetDirectDependencyNames` calls `ParseRootPackageTree` internally, what about an additional `ptree` return from `GetDirectDependencyNames`?
```
GetDirectDependencyNames(sm gps.SourceManager) (pkgtree.PackageTree, map[gps.ProjectRoot]bool, error) {
```

Reply:
Looks like that nil pointer is from the cache field too: ```if len(p.RootPackageTree.Packages)```

Reply:
i waffled over this a bit. it was between that signature, and what i did here - because we absolutely can't create a circumstance where casual use of the API results in hitting the disk twice for data that, at least in dep's model, hasn't changed.

however, i didn't particularly like having the extra return value - it felt like exposing a bit too much detail of the underlying behavior. the `ptree` isn't what we asked for here. but, now that i've finished writing it and looking back, i'd agree that the cost is less than the complexity arising from adding this new statefulness dimension to `*dep.Project`.

so yeah, i'll switch it around.

Reply:
actually looking at the code again, i've flipped back. not caching it internally also makes the `FindIneffectualConstraints()` method much more complicated to implement.

there are a lot of different possible paths through this code, and orders in which different use cases might call it. we can't reasonably take out the caching. but i suppose that also adding the additional return value doesn't hurt - at minimum, it's one less instance of error handling (which was your original concern).